repose-aggregator/components/filters/openstack-identity-v3/src/test/scala/org/openrepose/filters/openstackidentityv3/OpenStackIdentityV3HandlerTest.scala

```diff
@@ -159,6 +159,23 @@ class OpenStackIdentityV3HandlerTest extends FunSpec with BeforeAndAfter with Ma
       headers.keySet() should not contain HeaderName.wrap("X-Impersonator-Id")
     }
 
+    it("should non-destructively add the x-roles header") {
+      when(identityAPI.validateToken("123456", None)).thenReturn(
+        Try(new AuthenticateResponse("1", "2", List(), None, None, Option(List(ServiceForAuthenticationResponse(List(Endpoint("foo", None, None, None, "http://www.notreallyawebsite.com"))))),
+          Option(List(Role("1", "admin"))), UserForAuthenticateResponse(null, None, None, None, None, Some("ORD")))))
+      val mockRequest = new MockHttpServletRequest()
+      mockRequest.setHeader("X-Subject-Token", "123456")
+      identityConfig.setForwardGroups(false)
+      identityConfig.setValidateProjectIdInUri(null)
+      identityV3Handler = new OpenStackIdentityV3Handler(identityConfig, identityAPI)
+      identityV3Handler.handleRequest(mockRequest, mockServletResponse).requestHeaderManager.headersToAdd should contain(
+        Entry(
+          HeaderName.wrap(OpenStackServiceHeader.ROLES.toString),
+          JavaConversions.setAsJavaSet(Set("admin")))
+      )
+      identityV3Handler.handleRequest(mockRequest, mockServletResponse).requestHeaderManager.headersToRemove should not contain HeaderName.wrap(OpenStackServiceHeader.ROLES.toString)
```

Comment:
Wouldn't it be clearer if the `mockRequest` started with an `X-Roles` header that was not *admin* and then check to make sure it was still present and not marked for removal instead of calling `handleRequest()` twice to put the exact same header?